### PR TITLE
Handle PUT requests for pushing image signatures

### LINKED ISSUE #502: As a user I can publish container signature to the server
Author: @ipanova (ipanova@redhat.com)

Redmine Issue: 9511, https://pulp.plan.io/issues/9511

---

To add a single signature, PUT a new json object with `version` set to 2( manifest schema version) , `type` set to atomic, and `content` set to the base64 encoded signature data (usually a GPG signed data).
Name should be set to an unique name with the form digest@per-image-name, where digest is an image manifest digest (also used in the URL), and per-image-name is any unique identifier.
```
$ curl -X PUT --data @signature.json http://<registry_endpoint>:24817/extensions/v2/<namespace>/<name>/signatures/<imagesha256digest>

cat signature.json

{
    "version": 2,
    "type":    "atomic",
    "name":    "sha256:4028782c08eae4a8c9a28bf661c0a8d1c2fc8e19dbaae2b018b21011197e1484@cddeb7006d914716e2728000746a0b23",
    "content": "<base64 encoded signature>",
  }
```

See https://github.com/openshift/openshift-docs/pull/3556/files and https://docs.openshift.com/container-platform/3.10/admin_guide/image_signatures.html#writing-image-signatures-using-registry-api

**Note** skopeo will upload signature given that regitry has present `X-Registry-Supports-Signatures ` header  https://github.com/mtrmac/image/commit/6c17ca34793b19accc3d278fc93ce68e9943fcb4

### PR DESCRIPTION
**Pull request by lubosmj**
closes #502

### CODE REVIEW

File: CHANGES/502.feature [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Added support for pushing image signatures to the Pulp Registry. The signatures can be pushed by
```
Comment (ipanova) [inline CHANGES/502.feature:1]: can you extend existing docs? for example add another section to the signing workflow and mention another way how get signature into pulp registry.

File: pulp_container/app/registry_api.py [hunk 2]
```diff
@@ -234,6 +259,21 @@ def __init__(self, manifest, path, request, status=200):
...
+class ManifestSignatureResponse(Response):
+    """
+    An HTTP response class after creating an image signature.
+    """
+
+    def __init__(self, signature, path, status=201):
+        """Initialize the headers with the path to the repository and corresponding digests."""
+        headers = {
+            "Docker-Content-Digest": signature.digest,
+            "Location": f"/extensions/v2/{path}/signatures/{signature.signed_manifest.digest}",
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:271]: we have a mixture of string formatting, it'd be great to have one style :/

File: pulp_container/app/registry_api.py [hunk 3]
```diff
@@ -234,6 +259,21 @@ def __init__(self, manifest, path, request, status=200):
...
 
+class ManifestSignatureResponse(Response):
+    """
+    An HTTP response class after creating an image signature.
+    """
+
+    def __init__(self, signature, path, status=201):
+        """Initialize the headers with the path to the repository and corresponding digests."""
+        headers = {
+            "Docker-Content-Digest": signature.digest,
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:270]: it cannot be signature.digest. PUT can contain a list of signature for 1 image, if anything then it should be digest of the image but in case of docker-content-digest header - usually this header contains digest of the object that has been just created/fetched.
tldr; remove this header

File: pulp_container/app/registry_api.py [hunk 4]
```diff
@@ -1009,3 +1049,61 @@ def get_response_data(signatures):
...
+        """Create a new signature from the received data."""
+        _, repository = self.get_dr_push(request, path)
+
+        try:
+            manifest = models.Manifest.objects.get(digest=pk)
+        except models.Manifest.DoesNotExist:
+            raise ManifestNotFound(reference=pk)
+
+        signature_payload = request.META["wsgi.input"].read(SIGNATURE_PAYLOAD_MAX_SIZE)
+        signature_raw = base64.b64decode(json.loads(signature_payload)["content"])
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:1063]: what if the signature's payload will exceed the limit, will json be able to load and access the `content`?

File: pulp_container/app/registry_api.py [hunk 5]
```diff
@@ -1009,3 +1049,61 @@ def get_response_data(signatures):
...
+        signature_json = extract_data_from_signature(signature_raw, manifest.digest)
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:1064]: this utility accounts only for one signature extraction, PUT can contain 1+ signatures.

File: pulp_container/app/registry_api.py (hunk 4 above)
Comment (ipanova) [reply pulp_container/app/registry_api.py:1063]: we need to assert that body of PUT request is correct, namely that each signature instance contains schemaversion ==2 to and type == atomic. Name also should be required and i think it has some rules

File: pulp_container/app/registry_api.py [hunk 6]
```diff
@@ -1009,3 +1049,61 @@ def get_response_data(signatures):
...
+            type=SIGNATURE_TYPE.ATOMIC_SHORT,
+            key_id=signature_json["signing_key_id"],
+            timestamp=signature_json["signature_timestamp"],
+            creator=signature_json["optional"].get("creator"),
+            data=base64.b64encode(signature_raw).decode(),
+            signed_manifest=manifest,
+        )
+        try:
+            signature.save()
+        except IntegrityError:
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:1081]: reading this code it feels like integrity error should not happen because existing signatures are not getting (re)uploaded, but let's play safe https://github.com/containers/image/blob/main/docker/docker_image_dest.go#L654

File: pulp_container/app/registry_api.py [hunk 7]
```diff
@@ -1009,3 +1049,61 @@ def get_response_data(signatures):
...
+            signature = models.ManifestSignature.objects.get(name=signature.name)
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:1082]: i suggest get by digest and not name

File: pulp_container/app/registry_api.py [hunk 8]
```diff
@@ -1009,3 +1049,61 @@ def get_response_data(signatures):
             }
             data.append(signature)
         return {"signatures": data}
+
+    def put(self, request, path, pk):
+        """Create a new signature from the received data."""
+        _, repository = self.get_dr_push(request, path)
+
+        try:
+            manifest = models.Manifest.objects.get(digest=pk)
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:1058]: please also check that the manifest is in the repo

File: pulp_container/app/registry_api.py (hunk 6 above)
Comment (lubosmj) [reply pulp_container/app/registry_api.py:1081]: I will keep the try-except block.

File: pulp_container/app/registry_api.py (hunk 5 above)
Comment (lubosmj) [reply pulp_container/app/registry_api.py:1064]: Could you please provide a reference for this statement? I was looking into https://github.com/openshift/origin/pull/12504/files#diff-a215f9ef5344e2e492dbb2e3b82acf1d51f330e76e2c1dcc51b10a896e2d22fcR76 and https://docs.openshift.com/container-platform/3.11/admin_guide/image_signatures.html#writing-image-signatures-using-registry-api.

Comment (ipanova) [reply pulp_container/app/registry_api.py:1064]: I've been looking here https://github.com/mtrmac/image/commit/6c17ca34793b19accc3d278fc93ce68e9943fcb4#diff-6c682511ebec73d3ee49b3ed3d33708b9c598f9dd59c4f0685e793cc16baed8fR376
I do see discrepancies in different sources, yes.  let's discuss more on monday

Comment (ipanova) [reply pulp_container/app/registry_api.py:1064]: apparently the client parses the list of signatures and then sends N number of PUT requests  with one signature https://github.com/mtrmac/image/commit/6c17ca34793b19accc3d278fc93ce68e9943fcb4#diff-6c682511ebec73d3ee49b3ed3d33708b9c598f9dd59c4f0685e793cc16baed8fR408

Comment (lubosmj) [reply pulp_container/app/registry_api.py:1064]: Exactly. So, I will not update the code to handle this situation.

File: docs/workflows/sign-images.rst [hunk 9]
```diff
@@ -298,3 +298,27 @@ It is possible to specify a single manifest identified by tag or a list of manif
 by proviging ``tags_list`` option to the call.
 Note that ``manifest lists`` are not signed, instead all the image manifests that manifest lists
 contain, are signed.
+
+Push signed images to the Pulp Registry
```
Comment (ipanova) [inline docs/workflows/sign-images.rst:302]: push image signature to the pulp registry

File: docs/workflows/sign-images.rst [hunk 10]
```diff
@@ -298,3 +298,27 @@ It is possible to specify a single manifest identified by tag or a list of manif
...
+=======================================
+
+Signatures can be pushed to the Pulp Registry via the extensions API. This API exposes an endpoint
+for reading and writing image signatures.
+
+To add a new signature to an image, issue the following PUT request::
+
+    $ http PUT http://localhost:24817/extensions/v2/<namespace>/<name>/signatures/sha256:<manifest-digest> < signature.json
+
+The JSON payload should have the following structure::
```
Comment (ipanova) [inline docs/workflows/sign-images.rst:312]: how about referencing  for more info/specs to this source https://github.com/containers/image/blob/main/docs/signature-protocols.md#openshift-dockerdistribution-api-extension

File: docs/workflows/sign-images.rst [hunk 11]
```diff
@@ -298,3 +298,27 @@ It is possible to specify a single manifest identified by tag or a list of manif
...
+
+    {
+      "schemaVersion": 2,
+      "type":    "atomic",
+      "name":    "sha256:4028782c08eae4a8c9a28bf661c0a8d1c2fc8e19dbaae2b018b21011197e1484@cddeb7006d914716e2728000746a0b23",
+      "content": "<cryptographic_signature>"
+    }
+
+This step can be also done via podman. After configuring a GPG keyring, it is possible to issue
+the following command to push a tagged and signed manifest to the Pulp Registry::
```
Comment (ipanova) [inline docs/workflows/sign-images.rst:322]: ... to push a tagged image altogether with its signature.

File: docs/workflows/sign-images.rst [hunk 12]
```diff
@@ -298,3 +298,27 @@ It is possible to specify a single manifest identified by tag or a list of manif
...
+The JSON payload should have the following structure::
+
+    {
+      "schemaVersion": 2,
+      "type":    "atomic",
+      "name":    "sha256:4028782c08eae4a8c9a28bf661c0a8d1c2fc8e19dbaae2b018b21011197e1484@cddeb7006d914716e2728000746a0b23",
+      "content": "<cryptographic_signature>"
+    }
+
+This step can be also done via podman. After configuring a GPG keyring, it is possible to issue
```
Comment (ipanova) [inline docs/workflows/sign-images.rst:321]: ... via podman or skopeo, or any another client that supports registry extensions API for signatures.

File: docs/workflows/sign-images.rst [hunk 13]
```diff
@@ -298,3 +298,27 @@ It is possible to specify a single manifest identified by tag or a list of manif
 by proviging ``tags_list`` option to the call.
 Note that ``manifest lists`` are not signed, instead all the image manifests that manifest lists
 contain, are signed.
+
+Push signed images to the Pulp Registry
+=======================================
+
+Signatures can be pushed to the Pulp Registry via the extensions API. This API exposes an endpoint
```
Comment (ipanova) [inline docs/workflows/sign-images.rst:305]: can you mention that containers registries.d file sigstore section needs to be configured accordingly? or point to the specs https://github.com/containers/image/blob/main/docs/containers-registries.d.5.md

Comment (ipanova) [reply docs/workflows/sign-images.rst:305]: on the client side.

File: docs/workflows/sign-images.rst (hunk 9 above)
Comment (ipanova) [reply docs/workflows/sign-images.rst:302]: let me know what you think -not exactly sure this is the best place but it's convenient to add another section right below since it's about extensions api - 'serve signed images/image signature from the pulp registry.'
i'd cover in this section that  synced/pushed/or signed from within pulp signatures can be served by these extensions api and show an example of a GET call

Comment (ipanova) [reply docs/workflows/sign-images.rst:302]: and also a note about containers registries.d file sigstore configuration on the client side

File: pulp_container/app/registry_api.py [hunk 14]
```diff
@@ -169,6 +177,24 @@ def __init__(self, digest):
...
+
+    def __init__(self, digest):
+        """Initialize the exception with the digest of a signed manifest."""
+        super().__init__(
+            detail={
+                "errors": [
+                    {
+                        "code": "SIGNATURE_INVALID",
+                        "message": "signature invalid",
+                        "detail": {"digest": digest},
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:191]: how about manifest_digest, at first i thought this is signature digest until i looked further into the code.

File: pulp_container/app/registry_api.py [hunk 15]
```diff
@@ -1009,3 +1051,75 @@ def get_response_data(signatures):
...
+        except binascii.Error:
+            raise ManifestSignatureInvalid(digest=pk)
+
+        signature_json = extract_data_from_signature(signature_raw, manifest.digest)
+        if signature_json is None:
+            raise ManifestSignatureInvalid(digest=pk)
+
+        sig_digest = hashlib.sha256(signature_raw).hexdigest()
+        signature = models.ManifestSignature(
+            name=f"{manifest.digest}@{sig_digest[:32]}",
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:1086]: please do like here https://github.com/pulp/pulp_container/pull/533/files#diff-b702cbd3ebdff19bf8e4dfc14688aa3fb36e1a47ef6cf84d24e0f75b61247d36R365

File: pulp_container/app/registry_api.py [hunk 16]
```diff
@@ -1009,3 +1051,75 @@ def get_response_data(signatures):
...
+            digest=f"sha256:{sig_digest}",
+            type=SIGNATURE_TYPE.ATOMIC_SHORT,
+            key_id=signature_json["signing_key_id"],
+            timestamp=signature_json["signature_timestamp"],
+            creator=signature_json["optional"].get("creator"),
+            data=base64.b64encode(signature_raw).decode(),
```
Comment (ipanova) [inline pulp_container/app/registry_api.py:1092]: you can probably use directly here `signature_dict["content"]`

File: pulp_container/app/serializers.py [hunk 17]
```diff
@@ -160,6 +161,22 @@ class Meta:
...
 
 
+class ManifestSignaturePutSerializer(serializers.Serializer):
+    """
+    A serializer for image signatures provided in a PUT request.
+    """
+
+    name = serializers.RegexField(regex=VALID_SIGNATURE_NAME_REGEX)
+    schemaVersion = serializers.IntegerField(max_value=2, min_value=2)
+    type = serializers.CharField()
```
Comment (ipanova) [inline pulp_container/app/serializers.py:171]: you can make it a choice field with1 choice of atomic type

File: pulp_container/app/registry_api.py (hunk 15 above)
Comment (ipanova) [reply pulp_container/app/registry_api.py:1086]: i have not seen the serializer for this yet, so feel free to ignore my previous comment

File: pulp_container/app/serializers.py [hunk 18]
```diff
@@ -160,6 +161,17 @@ class Meta:
...
 
 
+class ManifestSignaturePutSerializer(serializers.Serializer):
+    """
+    A serializer for image signatures provided in a PUT request.
+    """
+
+    name = serializers.RegexField(regex=VALID_SIGNATURE_NAME_REGEX)
+    schemaVersion = serializers.IntegerField(max_value=2, min_value=2)
+    type = serializers.ChoiceField(["atomic"])
```
Comment (ipanova) [inline pulp_container/app/serializers.py:171]: we have a constant defined in constans.py you can use

**Review by ipanova — APPROVED** [review]
thank you!
